analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/RequestsWithExcessiveLengthBase.cs

```diff
@@ -59,28 +58,76 @@ protected RequestsWithExcessiveLengthBase(System.Resources.ResourceManager rspec
         }
 
         protected override void Initialize(ParameterLoadingAnalysisContext context) =>
-            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer,
+            context.RegisterCompilationStartAction(
                 c =>
                 {
-                    if (!IsEnabled(c.Options))
-                    {
-                        return;
-                    }
-
-                    if (c.Node is TAttributeSyntax attribute
-                        && (attribute.IsKnownType(KnownType.Microsoft_AspNetCore_Mvc_DisableRequestSizeLimitAttribute, c.SemanticModel)
-                            || IsInvalidRequestSizeLimit(attribute, c.SemanticModel)
-                            || IsInvalidRequestFormLimits(attribute, c.SemanticModel)))
-                    {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, c.Node.GetLocation()));
-                    }
-                },
-                Language.SyntaxKind.Attribute);
+                    var attributesOverTheLimit = new ConcurrentDictionary<SyntaxNode, Attributes>();
+
+                    c.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer,
+                        cc =>
+                        {
+                            if (!IsEnabled(c.Options))
+                            {
+                                return;
+                            }
+
+                            var attribute = (TAttributeSyntax)cc.Node;
+
+                            if (attribute.IsKnownType(KnownType.Microsoft_AspNetCore_Mvc_DisableRequestSizeLimitAttribute, cc.SemanticModel))
+                            {
+                                cc.ReportDiagnosticWhenActive(Diagnostic.Create(rule, attribute.GetLocation()));
+                            }
+
+                            if ((IsInvalidRequestSizeLimit(attribute, cc.SemanticModel)
+                                 || IsInvalidRequestFormLimits(attribute, cc.SemanticModel))
+                                && GetMethodOrClassDeclaration(attribute, cc.SemanticModel) is { } delcaration)
+                            {
+                                var isRequestFormLimits = attribute.IsKnownType(KnownType.Microsoft_AspNetCore_Mvc_RequestFormLimitsAttribute, cc.SemanticModel);
+
+                                attributesOverTheLimit.AddOrUpdate(
+                                    delcaration,
+                                    new Attributes
+                                    {
+                                        InvalidRequestForm = isRequestFormLimits ? attribute : null,
```

Comment:
It seems that `InvalidRequestForm` is exclusive with `InvalidRequestSize` - which means that one of the fields will always be null.  That makes the code hard to follow.

What I see you want to do is:
- collect all attributes of the sought types for a method
- iterate over them

Can't you just keep a collection of `TAttributeSyntax` per Declaration , make sure you don't add duplicates, and then use that? And like this, you get rid of the unnecessary `Attribute` struct.

Reply:
Or you could make the `Attribute` a bit smarter, and let it have the same fields, but let it deal with them inside, in the constructor.

And the getter should be "MainAttribute" and "SecondaryAttribute" which can be optional.

But the logic would be contained in the `Attribute` class.